<a href="https://colab.research.google.com/github/EshaanOPatel/MachineLearning/blob/main/CBBBasic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import Libraries

In [ ]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf
import seaborn as sns
from sklearn.metrics import mean_absolute_error

Load Data

In [ ]:
cbb_raw_data = pd.read_csv('/content/cbb.csv')
cbb_data = cbb_raw_data[['EFG_O','EFG_D','TOR','TORD','ORB','DRB','FTR','FTRD']]
y_train = cbb_raw_data.iloc[0:300, 3] 
x_train = cbb_data.iloc[0:300, :]
y_test = cbb_raw_data.iloc[300:354, 3]
x_test = cbb_data.iloc[300:354, :]
y_train.shape, x_train.shape, y_test.shape, x_test.shape

((300,), (300, 8), (54,), (54, 8))

Process Data: Normalize Features, Rescale Target

In [ ]:
# normalize features
x_train = (x_train - x_train.mean())/x_train.std()
x_test = (x_test - x_test.mean())/x_test.std()

# rescale target
#y_train = y_train/y_train.max()
#y_test = y_test/y_test.max()

##Model 1: Linear Regression 

In [ ]:
#Build Model
#model = tf.keras.models.Sequential()
#model.add(tf.keras.layers.Dense(1, input_shape=[8]))

##Model 2: Neural Network


In [ ]:
#Build Model
model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(256, input_shape=[8]))
model.add(tf.keras.layers.Dense(64, activation="relu"))
model.add(tf.keras.layers.Dense(16, activation="relu"))
model.add(tf.keras.layers.Dense(1))

##Testing


In [ ]:
#compile
model.compile(loss='mse', optimizer='sgd')

In [ ]:
#train model
model.fit(
    x=x_train, # features
    y=y_train, # target
    validation_split=0.2, #20% cross validation
    epochs=10
)

Epoch 1/10
8/8 [==============================] - 1s 22ms/step - loss: 227.6070 - val_loss: 90.0438
Epoch 2/10
8/8 [==============================] - 0s 5ms/step - loss: 286.2681 - val_loss: 417.7062
Epoch 3/10
8/8 [==============================] - 0s 6ms/step - loss: 429.0243 - val_loss: 129.6065
Epoch 4/10
8/8 [==============================] - 0s 5ms/step - loss: 240.5559 - val_loss: 90.4264
Epoch 5/10
8/8 [==============================] - 0s 5ms/step - loss: 195.1995 - val_loss: 66.7484
Epoch 6/10
8/8 [==============================] - 0s 6ms/step - loss: 160.8471 - val_loss: 49.4469
Epoch 7/10
8/8 [==============================] - 0s 6ms/step - loss: 134.7246 - val_loss: 37.4610
Epoch 8/10
8/8 [==============================] - 0s 6ms/step - loss: 115.5548 - val_loss: 30.3067
Epoch 9/10
8/8 [==============================] - 0s 5ms/step - loss: 102.2774 - val_loss: 26.1815
Epoch 10/10
8/8 [==============================] - 0s 5ms/step - loss: 92.7963 - val_loss: 23.9105


In [ ]:
#Evaluate model using Mean Absolute Error
prediction = model.predict(x_test)
print("MAE=")
mean_absolute_error(y_true=y_test, y_pred=prediction)

MAE=


3.7605868445502386

In [ ]:
print("MSE")
model.evaluate(x_test,y_test)

MSE
2/2 [==============================] - 0s 5ms/step - loss: 21.1232


21.123239517211914

Conclusion

| Model | Epochs |MSE |MAE|
|  ---  |  ---   | ---  | --- |
| Linear Regression | 10 |9.43|2.49|
|Linear Regression|50|17.46|3.43|
|[128,32]|10|10.99|2.57|
|[256,64]|10|40.09|5.24|
|[256,64,16]|10|21.12|3.76|